### PR TITLE
Support arbitrary properties in principal

### PR DESCRIPTION
**Pull request by GNSSYSTEMS-tg**
* added `data` field for `Principal`

This makes the principal more generic. This will be the last PR for `2.0`.

### CODE REVIEW

**Comment by IsNull** [comment]
The idea around the principal object is that it defines a clear, abstract API to interact with a principal. Instead of such a generic field, we probably should just allow access to the data behind it.

```
const data = principal.getProperty("mycustom")
```

The implementation could either redirect the property key to the claims, or we just read all properties from the JWT claims and store them in a map inside the principal. API wise, we wont expose the JWT.

If I may ask, what kind of property do you need? Maybe it would  be reasonable to have it as first class citizen anyway?

**Comment by GNSSYSTEMS-tg** [comment]
includes #17

**Comment by GNSSYSTEMS-tg** [comment]
I'm not sure yet, which information will finally make it the jwt. Our use case is the authentication via an Active Directory, so multiple additional information is available (e.g. full name, email address, office). I'd prefer to allow generic fields besides the standard. Your approach looks accurate, though.

**Comment by GNSSYSTEMS-tg** [comment]
* added `Principal.hasProperty(propertyName: string): any`
* added `Principal.getProperty(propertyName: string): any`
* added dependency to `lodash` (not yet documented)

File: src/auth/principal.ts [hunk 1]
```diff
@@ -15,20 +15,24 @@ export class Principal {
 
     private _realm: string;
 
+    private _data: any;
```
Comment (IsNull) [inline src/auth/principal.ts]: Please use a Map here. Probably `Map<string, string> properties`.

File: src/auth/services/token.service.ts [hunk 2]
```diff
@@ -257,7 +258,11 @@ export class TokenService {
             jwt[this.config.jwtRolesField],
             jwt[this.config.jwtIssuerField],
             jwt[this.config.jwtAudienceField],
-            jwt[this.config.jwtRealmField]
+            jwt[this.config.jwtRealmField],
+            _.omit(
+                jwt,
+                _.keys(this.config).filter(key => key.startsWith('jwt')),
+            ),
```
Comment (IsNull) [inline src/auth/services/token.service.ts]: I would prefer if we hold the number of dependencies down to a bare minimum in a base library as this. That being said, I don't think we have to necessary omit the already mapped fields.

By using TS features like `Object.keys(obj)` it should be easily possible without additional dependencies. :)

File: README.md [hunk 3]
```diff
@@ -271,6 +271,7 @@ export class JwtAuthConfig {
         issuerField?: string;
         audienceField?: string;
         realmField?: string;
+        dataField?: string;
```
Comment (IsNull) [inline README.md]: This can be remove I guess.

**Comment by GNSSYSTEMS-tg** [comment]
The JWT content is now accessible as whole, although I have to admit a reduced instance would be more handy to force the developer to stick to the standard.

File: src/auth/services/token.service.ts [hunk 4]
```diff
@@ -257,7 +258,8 @@ export class TokenService {
             jwt[this.config.jwtRolesField],
             jwt[this.config.jwtIssuerField],
             jwt[this.config.jwtAudienceField],
-            jwt[this.config.jwtRealmField]
+            jwt[this.config.jwtRealmField],
+            jwt,
```
Comment (IsNull) [inline src/auth/services/token.service.ts]: Assigning an object directly to a map seems a bit of a stretch? Does this even compile on your end?

Comment (GNSSYSTEMS-tg) [reply src/auth/services/token.service.ts]: I have to admit, I'm not quite sure, how the final solution is supposed to look like. As you described previously, having the complete jwt available via `getProperty` is desired modification.

Unfortunately I'm out of spare time for this. I'd like to have this feature, but I can't seem to find a proper solution in a short amount of time.

Comment (IsNull) [reply src/auth/services/token.service.ts]: No problem, I try to finish it if I find some time. 

>how the final solution is supposed to look like

Well its basically just enumerating all properties of the object and putting them into the map.

**Pull request merged**